fastparquet/util.py

```diff
@@ -1,22 +1,22 @@
+from collections import defaultdict
 import copy
+from distutils.version import LooseVersion
+from functools import lru_cache
 import io
 import struct
-import numpy as np
 import os
-import os.path
 import operator
-import pandas as pd
 import re
 import numbers
-from collections import defaultdict
-from distutils.version import LooseVersion
-from functools import lru_cache
-import pandas
 
+import numpy as np
+import pandas as pd
 from pandas.api.types import is_categorical_dtype
 
-PANDAS_VERSION = LooseVersion(pandas.__version__)
-created_by = "fastparquet-python version 1.0.0 (build 111)"
+from fastparquet import __version__
+
+PANDAS_VERSION = LooseVersion(pd.__version__)
+created_by = f"fastparquet-python version {__version__}"
```

Comment:
Please find a reference for this, I believe it is expected to have a very specific format (I didn't come up with "build xxx" myself!).

Reply:
We will need to update where we test if a dataset was created by fastparquet, which will now `startswith("fastparquet")`

Reply:
Well...
I made a quick search. This `created_by` variable appeared for the 1st time in [this commit](https://github.com/dask/fastparquet/commit/8f62b0efa195099228b6201afd64cc8981df0f52#diff-1d43052515aa054640353968f17622a524918efdb474e53d6ff483590ce2461b).

Before this variable (as visible in this same commit), the string being used was `'fastparquet-python'`, for instance:

![Screenshot from 2022-01-27 20-06-56](https://user-images.githubusercontent.com/13206058/151427022-76a763d2-230d-4a08-b1cb-a8fba66b4c42.png)

Do you have any idea? (I can guess this `build xxx` was possibly referring to the thrift stuff?

Reply:
At that point (Mar 21, 2017), things were still being ironed out. No, the "build xxx" is not tied to thrift, it's just a number because, presumably, I had found some place that indicated that such a number was required.

Reply:
Oh OK, that's fine. I thought (without checking) that it was an exact equals.

Reply:
> Oh OK, that's fine. I thought (without checking) that it was an exact equals.

Martin, to be exact, I made a thorough search and only found 3 occurences of this check, all being in `api.py`. I made a 2nd branch with this switch to `startswith`, you can check the comparison [here](https://github.com/dask/fastparquet/compare/main...yohplala:version2?expand=1).

The 1st occurence might be to be updated. It is a split, and I think either a `startswith` or a `in` would be better.

```python
# current code:
self.created_by.split(b' ', 1)[0] == b"fastparquet-python"
```

Reply:
What do you think?

Reply:
Let's go for `in` to be consistent

Reply:
This seems to be ok then! :)

Reply:
Can we please include the "build" part, since I feel it may be necessary?

Reply:
> Can we please include the "build" part, since I feel it may be necessary?

I am not opposed Martin, but to what number? Does it make sense to have a fixed number?
I just had a look to wikipedia, [software versioning](https://en.wikipedia.org/wiki/Software_versioning).
Looking for "build" word, here is the corresponding paragraphe:

![Screenshot from 2022-01-27 21-34-51](https://user-images.githubusercontent.com/13206058/151439362-106d4afb-1adc-42fb-832c-8722a856f406.png)

I stress the last part: _the definition of what constitutes a "major" as opposed to a "minor" change is entirely subjective and up to the author, as is what defines a "build"_

Re-using what is illustrated as `buil`, this could be the 3rd digit in 0.8.0.
So do you think it is ok to work out the `created_by` so that it looks like:
`'fastparquet-python version 0.8 build 0'` ?

Reply:
At least like "(build 0)", if you wish. 
https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L1028 suggests it should be a "hash" - maybe the git revision number? We don't use versioneer in this repo, we probably don't know the git hash without some work.

Reply:
> At least like "(build 0)", if you wish. https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L1028 suggests it should be a "hash" - maybe the git revision number? We don't use versioneer in this repo, we probably don't know the git hash without some work.

Ok, with the last commit, `created_by` now looks like this:
```python
from fastparquet.util import created_by
created_by
Out[1]: 'fastparquet-python version 0.8 (build 0)'
```

Please, let me know if this is ok.
Bests,

Reply:
Should be 0.8.0 ?

Reply:
> Should be 0.8.0 ?

I took the 3rd digit to make the build. This is both compliant with the article from Wikipedia:
_major.minor.build..._
and the impala extract you linked to:
_impala version 1.0 (build 6cf94d29b2b7115df4de2c06e2ab4326d721eb55)_

BUT,

I can also hardcode `(build 0)`, and leave `0.8.0` as it is in `__version__`.

Reply:
In most python usage, "version" is the whole string. We do *not* abide strictly by semantic-versioning in this or any other pyadata project I know of.